File: src/part02/chapter10/levelA/task12/java/logic/PolynomialHelper.java
package part02.chapter10.levelA.task12.java.logic;

import java.util.ArrayList;
import java.util.List;

public class PolynomialHelper {

    public List<Double> multiplyTwoPolynomial(List<Double> firstCoefficients, List<Double> secondCoefficients) {
        int resultPower = firstCoefficients.size() - 1 + secondCoefficients.size() - 1;
        if (firstCoefficients.size() > secondCoefficients.size()) {
            List<Double> temp = firstCoefficients;
            firstCoefficients = secondCoefficients;
            secondCoefficients = temp;
        }
        List<Double> result = new ArrayList<>(resultPower);
        for (int index = 0; index <= resultPower; index++) {
            double coefficient = 0.0;
            if (index == 0) {
                result.add(firstCoefficients.get(0) * secondCoefficients.get(0));
                continue;
            }
            if (index < firstCoefficients.size()) {
                for (int first = 0; first <= index; first++) {
                    coefficient += firstCoefficients.get(first) * secondCoefficients.get(index - first);
                }
            } else if (index >= firstCoefficients.size() && index < secondCoefficients.size()) {
                int first = 0;
                for (int second = index; first < firstCoefficients.size(); first++, second--) {
                    coefficient += firstCoefficients.get(first) * secondCoefficients.get(second);
                }
            } else if (index >= secondCoefficients.size()) {
                int second = secondCoefficients.size() - 1;
                for (int first = index - second; first < firstCoefficients.size() && second >= 0; first++, second--) {
                    coefficient += firstCoefficients.get(first) * secondCoefficients.get(second);
                }
            }
            result.add(index, coefficient);
        }
        return result;
    }

    public double obtainValueOfPolynomial(double variable, List<Double> coefficients) {
        double result = 0;
        for (int index = 0; index < coefficients.size(); index++) {
            result += Math.pow(variable, coefficients.size() - 1 - index) * coefficients.get(index);
        }
        return result;
    }

    public void printPolynomial(double variable, List<Double> coefficients) {
        for (int i = 0; i < coefficients.size(); i++) {
            if (i < coefficients.size() - 1) {
                System.out.println(coefficients.get(i) + " * (" + variable + ")^" + (coefficients.size() - i - 1) + "\t+");
            } else {
                System.out.println(coefficients.get(i) + " * (" + variable + ")^" + (coefficients.size() - i - 1));
            }
        }
    }
}

File: src/part02/chapter10/levelA/task12/kotlin/logic/PolynomialHelper.kt
package part02.chapter10.levelA.task12.kotlin.logic

import kotlin.math.pow

object PolynomialHelper {

    fun multiplyTwoPolynomials(firstCoefficients: List<Double>, secondCoefficients: List<Double>): List<Double> {
        val smallerBySizeCoefficients: List<Double>
        val biggerBySizeCoefficients: List<Double>
        if (firstCoefficients.size < secondCoefficients.size) {
            smallerBySizeCoefficients = firstCoefficients
            biggerBySizeCoefficients = secondCoefficients
        } else {
            smallerBySizeCoefficients = secondCoefficients
            biggerBySizeCoefficients = firstCoefficients
        }
        val resultPower = (smallerBySizeCoefficients.size - 1) + (biggerBySizeCoefficients.size - 1)
        val result = mutableListOf<Double>()
        for (index in 0..resultPower) {
            var currentCoefficient = 0.0
            when {
                index == 0 -> {
                    currentCoefficient = smallerBySizeCoefficients[0] * biggerBySizeCoefficients[0]

                }
                index < smallerBySizeCoefficients.size -> {
                    for (smallerCoefficientsIndex in 0..index) {
                        currentCoefficient +=
                                smallerBySizeCoefficients[smallerCoefficientsIndex] *
                                biggerBySizeCoefficients[index - smallerCoefficientsIndex]
                    }
                }
                index >= smallerBySizeCoefficients.size && index < biggerBySizeCoefficients.size -> {
                    var smallerCoefficientsIndex = 0
                    var biggerCoefficientsIndex = index
                    while (smallerCoefficientsIndex < smallerBySizeCoefficients.size) {
                        currentCoefficient +=
                                smallerBySizeCoefficients[smallerCoefficientsIndex] *
                                        biggerBySizeCoefficients[biggerCoefficientsIndex]
                        smallerCoefficientsIndex++
                        biggerCoefficientsIndex--
                    }
                }
                index >= biggerBySizeCoefficients.size -> {
                    var biggerCoefficientsIndex = biggerBySizeCoefficients.lastIndex
                    var smallerCoefficientsIndex = index - biggerCoefficientsIndex
                    while (smallerCoefficientsIndex < smallerBySizeCoefficients.size && biggerCoefficientsIndex >= 0) {
                        currentCoefficient +=
                                smallerBySizeCoefficients[smallerCoefficientsIndex] *
                                        biggerBySizeCoefficients[biggerCoefficientsIndex]
                        biggerCoefficientsIndex--
                        smallerCoefficientsIndex++
                    }
                }
            }
            result.add(index, currentCoefficient)
        }
        return result
    }

    fun obtainValueOfPolynomial(variable: Double, coefficients: List<Double>): Double {
        var result = 0.0
        for (index in coefficients.indices) {
            result += variable.pow(coefficients.lastIndex - index) * coefficients[index]
        }
        return result
    }

    fun printPolynomial(variable: Double, coefficients: List<Double>) {
        for (index in coefficients.indices) {
            if (index < coefficients.lastIndex) {
                println("${coefficients[index]} * ($variable)^${coefficients.lastIndex - index}\t+")
            } else {
                println("${coefficients[index]} * ($variable)^${coefficients.lastIndex - index}.")
            }
        }
    }
}